Make upload tool python3 compatible

mvdbeek opened the pull request:
I couldn't get make update_dependencies to run so that still needs to be done.

xref. #1715

mvdbeek committed:
Make upload tool python3 compatible

mvdbeek committed:
Consistently use webob.compat.cgi_FieldStorage

This works around bugs in python3 cgi.FieldStorage.

mvdbeek committed:
Fix batch api middleware

nsoranzo committed:
Remove WebOb pin in Pipfile

nsoranzo commented on lib/galaxy/web/framework/base.py:
```diff
@@ -346,7 +347,8 @@ def read_lines(self):
 
-cgi.FieldStorage = FieldStorage
+webob.compat.cgi_FieldStorage = FieldStorage
+webob.request.cgi_FieldStorage = FieldStorage
```
I don't think `webob.request.cgi_FieldStorage = FieldStorage` is needed:
```
>>> from webob.compat import cgi_FieldStorage as cgi_FieldStorage_compat
>>> from webob.request import cgi_FieldStorage as cgi_FieldStorage_request
>>> cgi_FieldStorage_compat == cgi_FieldStorage_request
True
>>> cgi_FieldStorage_request.__module__
'webob.compat'
```

mvdbeek replied:
It was needed when I tested it (this took me quite a while to figure out). I think this is because we import `webob.exc` which import from `webob.request` and  `cgi_FieldStorage` is imported in the [`webob.request` module](https://github.com/Pylons/webob/blob/master/src/webob/request.py#L38), and if not patching this explicitly is being used from the local scope in various places in https://github.com/Pylons/webob/blob/master/src/webob/request.py#L223. Does that make sense ?

mvdbeek replied:
You can test this by removing `webob.request.cgi_FieldStorage = FieldStorage` and running `sh run_tests.sh -api test/api/test_dataset_collections.py:DatasetCollectionApiTestCase.test_upload_collection`

nsoranzo replied:
I see! Then I think the best way to proceed would be to patch the `cgi_FieldStorage` class, instead of subclassing it and then replacing it, see https://tryolabs.com/blog/2013/07/05/run-time-method-patching-python/

I'll do that, test, and push to your branch, shout if you think that's not good!

nsoranzo commented:
There are 2 more instances of `cgi.FieldStorage`, if you want to change them too:
- `config/plugins/webhooks/demo/tour_generator/__init__.py`
- `lib/galaxy/webapps/tool_shed/api/repositories.py`

mvdbeek committed:
Inject `self` and `obj` into global scope

The list comprehension code in `transform_lines` gets its own local
scope, so just copying the local scope into the eval functions local
scope doesn't work.

mvdbeek committed:
Use webob.compat's FieldStorage in the remaining codebase

mvdbeek commented:
Perfectly fine, thanks @nsoranzo

On Mon, Jul 16, 2018, 12:38 PM Nicola Soranzo <notifications@github.com>
wrote:

> *@nsoranzo* commented on this pull request.
> ------------------------------
>
> In lib/galaxy/web/framework/base.py
> <https://github.com/galaxyproject/galaxy/pull/6467#discussion_r202651949>:
>
> > @@ -346,7 +347,8 @@ def read_lines(self):
>              self.read_lines_to_eof()
>
>
> -cgi.FieldStorage = FieldStorage
> +webob.compat.cgi_FieldStorage = FieldStorage
> +webob.request.cgi_FieldStorage = FieldStorage
>
> I see! Then I think the best way to proceed would be to patch the
> cgi_FieldStorage class, instead of subclassing it and then replacing it,
> see https://tryolabs.com/blog/2013/07/05/run-time-method-patching-python/
>
> I'll do that, test, and push to your branch, shout if you think that's not
> good!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/galaxyproject/galaxy/pull/6467#discussion_r202651949>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGfVpbF4FuMoE3R90OXlOcznibioqoWEks5uHHs5gaJpZM4VD6ke>
> .
>

nsoranzo committed:
Patch webob.compat.cgi_FieldStorage instead of overwrite it

nsoranzo committed:
Add py34-unit to default tox tests

nsoranzo commented:
Rebased to fix a merge conflict.

nsoranzo committed:
Use six.moves for urlparse import

jmchilton commented:
Wonderful, thanks a bunch!
